design/bgp-bfd.md

```diff
@@ -0,0 +1,164 @@
+# BFD enablement in BGP
+
+## Summary
+
+The purpose of this enhancement is to use enable [BFD](https://en.wikipedia.org/wiki/Bidirectional_Forwarding_Detection) to be used in conjunction with BGP.
+
+BFD will be available only in the upcoming [FRR based implementation](0001-frr.md).
+
+## Motivation
+
+BFD provides a quicker path failure detection than BGP, and using them together will allow users to provide a better
+service.
+
+BFD for BGP is supported by FRR [out of the box](http://docs.frrouting.org/en/latest/bfd.html#bgp-bfd-configuration).
+
+### Goals
+
+* Provide a way to establish BFD on BGP session.
+* Provide a way to configure the parameters of the BFD session associated with a BGP session.
+
+### Non-Goals
+
+* Implement BFD on the legacy BGP implementation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to declare a BGP session to be backed up by BFD.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to set the BFD parameters related to a BGP session.
+
+## Design Details
+
+### BFD Design proposal
+
+The idea is to leverage FRR to enable BFD on a BGP session.
+
+When declaring a BFD peer, all it takes to enable BFD is to add a `bfd` property on the neighbour:
+
+```yaml
+    neighbor <A.B.C.D|X:X::X:X|WORD> bfd profile BFDPROF
+```
+
+The proposal here is to define a bfd profile section in the config structure that looks like:
+
+```yaml
+    bfd-profiles:
+    - name: bfdprofile1
+      receive-interval: 150
+      transmit-interval: 150
+      detect-multiplier: 10
+      echo-receive-interval: 20
+      echo-transmit-interval: 20
+      echo-mode: true
+      passive-mode: true
+      minimum-ttl: 5
+```
+
+When a property of the profile is not set, MetalLB will honor [FRR default values](https://docs.frrouting.org/en/latest/bfd.html#peer-profile-configuration).
+
+When setting a BGP peer, an optional bfd-profile property will enable BFD:
+
+```yaml
+    peers:
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      bfd-profile: bfdprofile1
+```
+
+#### Impacts on the operator
+
+Keeping the operator as the guinea pig for the CRD implementation, a new BFDProfile CRD will be introduced, with the form of:
```

Comment:
You don't actually have to do this, but as a suggestion for the future, I'd mark names of CRDs as literals like this:

```suggestion
Keeping the operator as the guinea pig for the CRD implementation, a new `BFDProfile` CRD will be introduced, with the form of:
```